doc_indexer/src/indexing/adaptive_indexer.py

```diff
@@ -0,0 +1,313 @@
+        return match.group(1).strip()
+    return None
+
+
+class AdaptiveSplitMarkdownIndexer:
+    """
+    Markdown indexer that adaptively splits documents based on token size thresholds,
+    using hierarchical headers (H1->H2->H3) only when needed until H3 level is reached.
+    """
+
+    def __init__(
+        self,
+        docs_path: str,
+        embedding: Embeddings,
+        connection: dbapi.Connection,
+        table_name: str | None = None,
+        headers_to_split_on: list[tuple[str, str]] | None = None,
+        min_chunk_token_count: int = 20,
+        max_chunk_token_count: int = 1000,
+    ):
+        self.headers_to_split_on = headers_to_split_on or [HEADER1, HEADER2, HEADER3]
+        if not table_name:
+            table_name = docs_path.split("/")[-1]
+
+        self.docs_path = docs_path
+        self.table_name = table_name
+        self.embedding = embedding
+        self.min_chunk_token_count = min_chunk_token_count
+        self.max_chunk_token_count = max_chunk_token_count
+
+        self.db = HanaDB(
+            connection=connection,
+            embedding=embedding,
+            table_name=table_name,
+        )
+
+        self.markdown_splitter_h1 = MarkdownHeaderTextSplitter(
+            headers_to_split_on=[HEADER1]
+        )
+
+        self.markdown_splitter_h2 = MarkdownHeaderTextSplitter(
+            headers_to_split_on=[HEADER1, HEADER2]
+        )
+
+        self.markdown_splitter_h3 = MarkdownHeaderTextSplitter(
+            headers_to_split_on=[HEADER1, HEADER2, HEADER3]
+        )
+
+    def _build_title(self, doc: Document) -> str:
+        # the following lines build the combined title from the headers H1, H2, H3
+        header1 = remove_header_brackets(doc.metadata.get("Header1", "")).strip()
+        header2 = remove_header_brackets(doc.metadata.get("Header2", "")).strip()
+        header3 = remove_header_brackets(doc.metadata.get("Header3", "")).strip()
+
+        # Join non-empty headers with " - "
+        title_parts = [part for part in [header1, header2, header3] if part]
+        title = " - ".join(title_parts)
+
+        return title
+
+    def _process_doc(
+        self,
+        doc: Document,
+        level: int = 0,
+        parent_title: str = "",
+    ) -> Generator[Document, None, None]:
+        tokens = len(encoding.encode(doc.page_content))
+
+        if tokens <= self.min_chunk_token_count:
+            return
+
+        # If the document is smaller than the max chunk token count or the H3 level is reached, yield the document
+        if tokens <= self.max_chunk_token_count or level >= len(HEADER_LEVELS):
+            yield Document(
+                page_content=doc.page_content,
+                metadata={
+                    "source": doc.metadata.get("source", ""),
+                    "title": doc.metadata.get("title")
+                    or extract_first_title(doc.page_content),
+                    "module": "Kyma Module",
+                    "version": "1.2.1",
```

Comment:
Why the version is hard coded?

Reply:
Should I add latest for now? We agreed to use the latest release for now.

Reply:
I've make it configurable for the function. We can use it later when we introduce module versions.